sklearn/ensemble/tests/test_iforest.py

```diff
@@ -235,19 +235,22 @@ def test_iforest_subsampled_features():
 
 
 def test_iforest_average_path_length():
+    def harmonic_humber(n):
+        return np.sum(1.0/np.arange(1, n+1))
     # It tests non-regression for #8549 which used the wrong formula
     # for average path length, strictly for the integer case
     # Updated to check average path length when input is <= 2 (issue #11839)
-    result_one = 2.0 * (np.log(4.0) + np.euler_gamma) - 2.0 * 4.0 / 5.0
-    result_two = 2.0 * (np.log(998.0) + np.euler_gamma) - 2.0 * 998.0 / 999.0
+    result_one = 2.0 * harmonic_humber(4.0) - 2.0 * 4.0 / 5.0
+    result_two = 2.0 * harmonic_humber(998.0) - 2.0 * 998.0 / 999.0
```

Comment:
I think we should leave the approximation with euler_gamma, since using the brute formula forces us to loosen the check tolerance later.

Also this isn't related to the changes (or only marginally)  as far as I can tell?

Reply:
I've checked the approximation precision separately.

My proposed approximation equation is closer to 'brute force' exact formula, than the current equation.

So I propose to keep this commit, and then I'll decrease the tolerance in this test in the following commit. In order to have a baseline, if somebody will find even more precise approximation in future.

Reply:
We  typically don't trust checks with a 0.1 rtol.

> My proposed approximation equation is closer to 'brute force' exact formula, than the current equation.

You haven't changed the approximation of H(n), you have only changed the approximtion of c(n). So we shouldn't change the approximation of H(n) in this check either.

Reply:
I am sorry, you are most probably right, but I cannot get your logic. First of all, we have an equation for `c(n)` written in terms of `H(n)`. This equation is ground truth one, if we follow exact `H(n)` definition. Without any code, for any `n`, we can evaluate real `c(n)` value on a paper.

So, my statement was that `2 (log(n) + gamma - 1)` produces the value which is closer to the real `c(n)` than the value produced by `2 (log(n-1) + gamma) - 2 (n - 1) / n`. It can be demonstrated numerically. So, here we have an improvement in the accuracy.

So why should we compare two approximations in the unit test?
`0.1` is a real relative tolerance for approximating real `c(n)` value by `2 (log(n-1) + gamma) - 2 (n - 1) / n` for `n=5`, nothing can be done here.

Reply:
Let's say that on master and in the paper, `c(n) = f(Hn)`

What you did is rewrite/simplify `c(n)` so it can be `c(n) = g(Hn)` where `g(Hn) == f(Hn)`.

The way `Hn` is computed is orthogonal to this change.
Also, in master *and* in this PR, `Hn` is approximated with the euler gamma.

What you're currently testing in this PR is that `g(approx(Hn)) == f(Hn)`.
If you leave this test as it is on master, the test would be `g(approx(Hn)) == f(approx(Hn))`, and I think that's a more sensible test.

If you wish, we can also add a test `g(approx(Hn)) == g(Hn)`

Anyway, it's not super important

Reply:
@NicolasHug Thank you for the explanation. However, I still tend to disagree a little.

I agree that the following is hold: `c(n) = g(Hn,n) = f(H_{n-1},n)`, where

```
g(Hn,n)       = 2 (Hn-1) = c(n)
f(H_{n-1}, n) = 2 H_{n-1} - 2 (n-1) / n = c(n)
```

Let `Hn = approx(Hn) + e(n)`, where `e(n)` is an approximation error for `n`-th Harmonic number, and then `approx(Hn) = Hn - e(n)`.

Then we have the following by the direct substitution:
```
g(approx(Hn), n) = c(n) - 2 e(n)
f(approx(H_{n-1}), n) = c(n) - 2 e(n-1)
```

So `g(approx(Hn),n) =\= f(approx(H_{n-1}),n)`.

Reply:
It's very simple:

- you did not change the way H is approximated in the code.
- So you shouldn't change the way it is approximated in the test.

The fact that you're using H(n) instead of H(n - 1) isn't relevant. I understand it will lead to a smaller error, but that smaller error is in the approximation of H, not directly in the approximtion of c.

Anyway, like I said this isn't important. resolving.